File: midi_lab/visualizer/gl/particles.py
# -*- coding: utf-8 -*-
"""GPU 向け大規模パーティクルプール — 鍵盤ヒットで上方向に飛散。"""
from __future__ import annotations

import random

import numpy as np

MAX_PARTICLES = 16384
PARTICLE_DTYPE = np.dtype(
    [
        ("px", "f4"),
        ("py", "f4"),
        ("vx", "f4"),
        ("vy", "f4"),
        ("life", "f4"),
        ("max_life", "f4"),
        ("size", "f4"),
        ("r", "f4"),
        ("g", "f4"),
        ("b", "f4"),
        ("a", "f4"),
    ]
)


class GPUParticleSystem:
    def __init__(self) -> None:
        self._data = np.zeros(MAX_PARTICLES, dtype=PARTICLE_DTYPE)
        self._count = 0
        self.amount_scale = 1.0

    def clear(self) -> None:
        self._count = 0

    def spawn_hit(
        self,
        x: float,
        y: float,
        velocity: int,
        *,
        r: float,
        g: float,
        b: float,
        style_kind: int = 0,
    ) -> None:
        vel_f = max(0.05, velocity / 127.0)
        n_spawn = max(8, int(32 * self.amount_scale * vel_f))
        for _ in range(n_spawn):
            if self._count >= MAX_PARTICLES:
                self._data = np.roll(self._data, -1, axis=0)
                self._count = MAX_PARTICLES - 1
            i = self._count
            self._count += 1
            spread = random.uniform(-0.55, 0.55)
            spd = random.uniform(120.0, 420.0) * vel_f
            life = random.uniform(0.4, 1.2)
            p = self._data[i]
            p["px"] = x + random.uniform(-6.0, 6.0)
            p["py"] = y
            if style_kind == 2:
                p["vx"] = spd * spread * 0.4
                p["vy"] = -spd * random.uniform(0.5, 1.0)
            elif style_kind == 3:
                p["vx"] = spd * spread
                p["vy"] = -spd * random.uniform(0.3, 0.7)
            else:
                p["vx"] = spd * spread * 0.65
                p["vy"] = -spd * random.uniform(0.75, 1.15)
            p["life"] = life
            p["max_life"] = life
            p["size"] = random.uniform(4.0, 16.0) * (0.5 + vel_f)
            p["r"] = r
            p["g"] = g
            p["b"] = b
            p["a"] = 1.0

    def update(self, dt: float) -> None:
        alive = 0
        for i in range(self._count):
            p = self._data[i]
            p["life"] -= dt
            if p["life"] <= 0:
                continue
            p["px"] += p["vx"] * dt
            p["py"] += p["vy"] * dt
            p["vy"] += 220.0 * dt
            p["vx"] *= 1.0 - 1.8 * dt
            if alive != i:
                self._data[alive] = p
            alive += 1
        self._count = alive

    def active_array(self) -> np.ndarray:
        return self._data[: self._count]

    def __len__(self) -> int:
        return self._count

File: rust/midi_viz/src/render/particles.rs
//! パーティクルプール（CPU スポーン + GPU シミュレーション）

use bytemuck::{Pod, Zeroable};

pub const MAX_PARTICLES: usize = 16384;

#[repr(C)]
#[derive(Clone, Copy, Pod, Zeroable)]
pub struct GpuParticle {
    pub pos: [f32; 2],
    pub vel: [f32; 2],
    pub life: f32,
    pub max_life: f32,
    pub size: f32,
    pub midi: f32,
    pub velocity: f32,
    pub _pad: f32,
}

pub struct ParticlePool {
    pub data: Vec<GpuParticle>,
    pub alive: usize,
    pub amount_scale: f32,
}

impl Default for ParticlePool {
    fn default() -> Self {
        Self {
            data: vec![GpuParticle::zeroed(); MAX_PARTICLES],
            alive: 0,
            amount_scale: 1.0,
        }
    }
}

impl ParticlePool {
    pub fn clear(&mut self) {
        self.alive = 0;
    }

    pub fn spawn_hit(&mut self, px: f32, py: f32, velocity: u8, midi: u8) {
        let n = (32.0 * self.amount_scale * (velocity as f32 / 127.0)).max(8.0) as usize;
        for _ in 0..n {
            if self.alive >= MAX_PARTICLES {
                self.alive = MAX_PARTICLES - 1;
            }
            let i = self.alive;
            self.alive += 1;
            let spread = (rand_simple() - 0.5) * 1.1;
            let spd = 120.0 + rand_simple() * 300.0;
            let vel_f = velocity as f32 / 127.0;
            let life = 0.4 + rand_simple() * 0.8;
            self.data[i] = GpuParticle {
                pos: [px, py],
                vel: [spread * spd * 0.65, -spd * (0.75 + rand_simple() * 0.4)],
                life,
                max_life: life,
                size: 4.0 + rand_simple() * 12.0 * vel_f,
                midi: midi as f32,
                velocity: velocity as f32,
                _pad: 0.0,
            };
        }
    }

    pub fn tick_cpu(&mut self, dt: f32) {
        let mut write = 0;
        for i in 0..self.alive {
            let mut p = self.data[i];
            p.life -= dt;
            if p.life <= 0.0 {
                continue;
            }
            p.vel[1] -= 220.0 * dt;
            p.vel[0] *= 1.0 - 1.8 * dt;
            p.pos[0] += p.vel[0] * dt;
            p.pos[1] += p.vel[1] * dt;
            self.data[write] = p;
            write += 1;
        }
        self.alive = write;
    }
}

fn rand_simple() -> f32 {
    use std::cell::Cell;
    thread_local! {
        static S: Cell<u32> = const { Cell::new(0x12345678) };
    }
    let mut x = S.get();
    x ^= x << 13;
    x ^= x >> 17;
    x ^= x << 5;
    S.set(x);
    (x & 0x00ff_ffff) as f32 / 16777216.0
}
